### PR TITLE
Split off ruff_cli crate from ruff library

### PR DESCRIPTION
**Pull request by not-my-profile**
Please preserve the commits when merging.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Nice! This will be a big improvement.

Before I dig in, I just want to confirm: there are intended to be no behavior changes here, right? As in, Ruff itself should function identically? Purely an internal refactor?

**Comment by not-my-profile** [comment]
Yes exactly.

File: Cargo.toml
Comment (charliermarsh) [inline Cargo.toml:7]: I have some thoughts, but in your opinion, how does this compare to the following alternative?

> Move everything in `src` into `ruff_core`, keep `ruff_cli` as it is (roughly), then change the root to point to `ruff_cli/src/main.rs` or appropriate.

This is, IIUC, how [`ripgrep`](https://github.com/BurntSushi/ripgrep) is structured.

Right now, there's a slight oddity here in that the top-level project isn't what's being packaged and distributed (which relates to having to change the working directory on CI).

File: ruff_cli/pyproject.toml
Comment (charliermarsh) [inline ruff_cli/pyproject.toml:16]: The `tool.setuptools` references below would need to be changed as well (though I think they're currently being ignored -- please change them though).

File: .github/workflows/ruff.yaml
Comment (messense) [reply .github/workflows/ruff.yaml:37]: Should work for building wheels but will have trouble with sdist, see https://github.com/charliermarsh/ruff/pull/1816#discussion_r1068265123, if you move `pyproject.toml` back to top level and set `manifest-path` option in it then you don't need this `working-directory` option.

File: .github/workflows/ruff.yaml
Comment (not-my-profile) [reply .github/workflows/ruff.yaml:17]: Ah yes, good catch!

File: ruff_cli/Cargo.toml
Comment (messense) [inline ruff_cli/Cargo.toml:32]: You can use cargo workspace inheritance feature to manage common dependencies between workspace members, see https://doc.rust-lang.org/cargo/reference/workspaces.html#the-dependencies-table

File: Cargo.toml
Comment (charliermarsh) [reply Cargo.toml:7]: (It would also be ok to say "We want to do this later, since it's another big change, but it is planned." I'm mostly trying to understand and align on the desired end-state.)

File: ruff_cli/pyproject.toml
Comment (not-my-profile) [reply ruff_cli/pyproject.toml]: Thanks, I made the changes you suggested :)

**Comment by messense** [comment]
I just checked locally, looks like maturin is having trouble with this kind of layout, mostly related to `python-source` option:

* `python/ruff/*` isn't in the right place in sdist
* `python/ruff/*` isn't packaged in wheel

I'll take a look tomorrow.

File: Cargo.toml
Comment (not-my-profile) [reply Cargo.toml:7]: I consider the "core" of ruff to be conceptually separate from the rule implementations, even if they are currently entangled. So I think the crate of the currently top-level `src` directory should keep the name `ruff`, as long as it contains both the core and the rule implementations.

While I think it makes some sense that the top-level `src` directory contains the code that contributors are most likely interested in editing, I'd be alright with moving `src` to `ruff/src` for the sake of consistency, however we might firstly want to think about further splitting up the crate, see #1820.

File: ruff_cli/Cargo.toml
Comment (not-my-profile) [reply ruff_cli/Cargo.toml]: Oh neat, I didn't know about that!

I guess you meant specifically for these verbose dependencies on `rustpython-*` right? I just changed `ruff::message` to `pub use rustpython_ast::Location;` and dropped the direct dependencies on rustpython-* from `ruff_cli`, which I think is even better :)

**Comment by charliermarsh** [comment]
@not-my-profile - How big of a problem will it be for you if I make some changes to the `pyproject.toml` resolution code? (Specifically, to solve #1812.) I'm happy to wait until this is merged if it saves you a headache, though guessing that will be tomorrow or the next day and not today given the packaging needs.

**Comment by not-my-profile** [comment]
Thanks for asking :)

You could merge the first two commits from this PR first (I just opened #1822 for that). Then you can change the resolution code without causing headaches :) (The remaining two commits of this PR can be easily rebased.)

**Comment by charliermarsh** [comment]
@not-my-profile - Hope you feel empowered to keep pushing on these sorts of refactors. They're making the project much stronger.

File: ruff_cli/Cargo.toml
Comment (messense) [inline ruff_cli/Cargo.toml:6]:
```suggestion
# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[[bin]]
name = "ruff"
path = "src/main.rs"
doctest = false
```

To keep the binary name `ruff`.

File: ruff_cli/Cargo.toml
Comment (messense) [inline ruff_cli/Cargo.toml:46]:
```suggestion
update-informer = ["dep:update-informer"]

[package.metadata.maturin]
name = "ruff"
```

To keep the python package name `ruff` after https://github.com/PyO3/maturin/pull/1409 lands.

**Comment by not-my-profile** [comment]
Thanks @messense I have applied your suggested changes locally ... I'll force push this once the new maturin version has been published so that I can also bump the minimum version in `pyproject.toml`:

```toml
[build-system]
requires = ["maturin>=0.14,<0.15"]
```

**Comment by messense** [comment]
@not-my-profile maturin [v0.14.10](https://pypi.org/project/maturin/0.14.10/) is out.

**Comment by charliermarsh** [comment]
Will review (and hopefully merge) later today.

**Comment by charliermarsh** [comment]
@not-my-profile - Can I squash the Clippy commit?

File: ruff_cli/Cargo.toml
Comment (charliermarsh) [inline ruff_cli/Cargo.toml:10]: @messense - This was also necessary. As-is, `ruff-0.0.220.dist-info/METADATA` was missing the README contents.

**Comment by charliermarsh** [comment]
Okay, the packages are looking more similar now:

Before:

```
❯ unzip -l /Users/crmarsh/workspace/ruff/target/wheels/ruff-0.0.220-py3-none-macosx_11_0_arm64.whl
Archive:  /Users/crmarsh/workspace/ruff/target/wheels/ruff-0.0.220-py3-none-macosx_11_0_arm64.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
   120482  01-14-2023 01:34   ruff-0.0.220.dist-info/METADATA
      103  01-14-2023 01:34   ruff-0.0.220.dist-info/WHEEL
     1070  01-14-2023 01:34   ruff-0.0.220.dist-info/license_files/LICENSE
        0  01-14-2023 01:34   ruff/__init__.py
      193  01-14-2023 01:34   ruff/__main__.py
 40412178  01-14-2023 01:34   ruff-0.0.220.data/scripts/ruff
      540  01-14-2023 01:34   ruff-0.0.220.dist-info/RECORD
```

After:
```
❯ unzip -l /Users/crmarsh/workspace/ruff/target/wheels/ruff-0.0.220-py3-none-macosx_11_0_arm64.whl
Archive:  /Users/crmarsh/workspace/ruff/target/wheels/ruff-0.0.220-py3-none-macosx_11_0_arm64.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
   120119  01-14-2023 01:34   ruff-0.0.220.dist-info/METADATA
      103  01-14-2023 01:34   ruff-0.0.220.dist-info/WHEEL
     1070  01-14-2023 01:34   ruff-0.0.220.dist-info/license_files/LICENSE
        0  01-14-2023 01:34   ruff/__init__.py
      193  01-14-2023 01:34   ruff/__main__.py
 38692434  01-14-2023 01:34   ruff-0.0.220.data/scripts/ruff
      540  01-14-2023 01:34   ruff-0.0.220.dist-info/RECORD
```

**Comment by charliermarsh** [comment]
@not-my-profile - In fact, can we squash this to a single commit, with the message you want?

**Comment by not-my-profile** [comment]
Since the first commit doesn't edit the files of the second commit, sure ... squashed into one :)

**Review by messense — APPROVED** [review]

**Comment by charliermarsh** [comment]
Awesome - thank you both!

**Comment by not-my-profile** [comment]
(I missed that we also have to update a command in the `playground.yml` workflow ... however `wasm-pack build` apparently doesn't support the new structure anyway ... I just opened #1860 to track that.)

**Comment by charliermarsh** [comment]
For reasons I don't fully understand, I think this change broke GitHub's dependency tracking:

Screen Shot 2023-01-14 at 3 00 48 PM

**Comment by charliermarsh** [comment]
I'd really love to get this back. I don't understand why it broke as we still have the top-level `setup.py`.

**Comment by not-my-profile** [comment]
Yeah I have also noticed that it's bugged ... when I refresh the page a couple of times it sometimes shows:

image

**Comment by charliermarsh** [comment]
Oh, wait, I see that now too...

**Comment by not-my-profile** [comment]
I don't think this PR changed any files that GitHub uses for dependency tracking ... maybe GitHub just cannot handle how many projects depend on ruff :P

**Comment by charliermarsh** [comment]
Interestingly, I now see this, which toggles between the results. Could we rename the Rust crate to avoid this issue? Would that work?

Screen Shot 2023-01-15 at 2 32 14 AM

**Comment by charliermarsh** [comment]
Ah, there's a setting for this! Ok, perfect.

Screen Shot 2023-01-15 at 2 37 01 AM

**Pull request merged**
